File: sentil-java/src/main/java/io/github/sedislab/sentil/StochasticSystem.java
package io.github.sedislab.sentil;

import java.util.Arrays;
import java.util.List;

/** A sampling-ready stochastic system, the form the rare-event estimator consumes. */
public final class StochasticSystem extends NativeResource {
    private final long box;

    StochasticSystem(long handle) {
        super(handle, NativeLib::stochasticSystemDestroy);
        this.box = 0L;
    }

    private StochasticSystem(long handle, long box) {
        super(handle, h -> {
            NativeLib.stochasticSystemDestroy(h);
            NativeLib.freeSystemBox(box);
        });
        this.box = box;
    }

    /** A system whose dynamics are host callbacks. They must be thread-safe. */
    public static StochasticSystem custom(String[] variables, double dt, long horizon,
            SystemInit init, SystemStep step) throws SentilException {
        long[] result = NativeLib.stochasticSystemCreateCustom(variables, dt, horizon, init, step);
        return new StochasticSystem(result[0], result[1]);
    }

    /** Simulate one full-horizon trajectory from a seed. */
    public Trace simulate(long seed) throws SentilException {
        Trace trace = new Trace(NativeLib.stochasticSystemSimulate(handle(), seed));
        try {
            rethrowCallbackError();
        } catch (RuntimeException e) {
            trace.close();
            throw e;
        }
        return trace;
    }

    /** The state variable names. */
    public List<String> variables() throws SentilException {
        return Arrays.asList(NativeLib.stochasticSystemVariables(handle()));
    }

    /** The time step. */
    public double dt() {
        return NativeLib.stochasticSystemDt(handle());
    }

    /** The number of steps in a trajectory. */
    public long horizon() {
        return NativeLib.stochasticSystemHorizon(handle());
    }

    void rethrowCallbackError() {
        if (box != 0L) {
            NativeLib.rethrowSystemError(box);
        }
    }
}

File: sentil-matlab/+sentil/StochasticSystem.m
classdef StochasticSystem < handle
    %STOCHASTICSYSTEM A sampling-ready stochastic system for rare-event estimation.

    properties (SetAccess = private, Hidden)
        Handle uint64 = uint64(0)
        Box uint64 = uint64(0)
    end

    methods (Static)
        function s = custom(variables, dt, horizon, init, step)
            %CUSTOM A system driven by init(seed) and step(prev, time, seed).
            if ~iscell(variables)
                variables = cellstr(variables);
            end
            [handle, box] = sentil_mex('stochastic_system_create_custom', init, step, ...
                variables, double(dt), double(horizon));
            s = sentil.StochasticSystem(handle, box);
        end
    end

    methods
        function obj = StochasticSystem(handle, box)
            obj.Handle = handle;
            if nargin >= 2
                obj.Box = box;
            end
        end

        function delete(obj)
            if obj.Handle ~= 0
                sentil_mex('stochastic_system_destroy', obj.Handle);
                obj.Handle = uint64(0);
            end
            if obj.Box ~= 0
                sentil_mex('free_system_box', obj.Box);
                obj.Box = uint64(0);
            end
        end

        function t = simulate(obj, seed)
            %SIMULATE One full-horizon trajectory from a seed.
            obj.assertOpen();
            t = sentil.Trace(uint64(sentil_mex('stochastic_system_simulate', obj.Handle, ...
                obj.Box, double(seed))));
        end

        function v = variables(obj)
            obj.assertOpen();
            v = sentil_mex('stochastic_system_variables', obj.Handle);
        end

        function d = dt(obj)
            obj.assertOpen();
            d = sentil_mex('stochastic_system_dt', obj.Handle);
        end

        function h = horizon(obj)
            obj.assertOpen();
            h = sentil_mex('stochastic_system_horizon', obj.Handle);
        end
    end

    methods (Access = private)
        function assertOpen(obj)
            if obj.Handle == 0
                error('sentil:handle', 'this system has been closed');
            end
        end
    end
end